### PR TITLE
MOD-5902: Report memory of the TAG TrieMaps and TEXT Tries

### PR DESCRIPTION
**Pull request by raz-mon**
**Describe the changes in the pull request**

This PR enhances the memory reporting of RediSearch, by reporting the overhead of the TAG TrieMaps and TEXT Tries that are not reported today.

A continuation PR should be made, containing the counting of the memory allocated for the values of the TAG `suffix` TrieMap and the TEXT `spec->suffix` Trie on-the-fly, i.e., when allocating and freeing them. This is in order to keep the `FT.INFO` command `O(1)`.

Other than that, there are some cleanups, and one revert of a fix that was applied to TEXT **and** TAG, where it was relevant only for TEXT.

We also add a locking of the `FT.INFO` section that reads fields that are exposed under the spec-lock (to avoid an incoherent response).

Has the foundations for MOD-6839 as well.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/spec.c [hunk 1]
```diff
@@ -2187,6 +2228,9 @@ void IndexSpec_AddToInfo(RedisModuleInfoCtx *ctx, IndexSpec *sp) {
   RedisModule_InfoAddFieldDouble(ctx, "doc_table_size", sp->docs.memsize / (float)0x100000);
   RedisModule_InfoAddFieldDouble(ctx, "sortable_values_size", sp->docs.sortablesSize / (float)0x100000);
   RedisModule_InfoAddFieldDouble(ctx, "key_table_size", TrieMap_MemUsage(sp->docs.dim.tm) / (float)0x100000);
+  RedisModule_InfoAddFieldDouble("tag_overhead_size_mb", IndexSpec_collect_tags_overhead(sp) / (float)0x100000);
+  RedisModule_InfoAddFieldDouble("text_overhead_size_mb", IndexSpec_collect_text_overhead(sp) / (float)0x100000);
+  RedisModule_InfoAddFieldDouble("total_index_memory_sz_mb", IndexSpec_TotalMemUsage(sp) / (float)0x100000);
```
Comment (raz-mon) [inline src/spec.c:2233]: Duplication of memory computation since we intend to remove the components of this field from this command in the near future.

File: src/spec.c [hunk 2]
```diff
@@ -2338,7 +2381,7 @@ int IndexSpec_CreateFromRdb(RedisModuleCtx *ctx, RedisModuleIO *rdb, int encver,
     if (FieldSpec_IsSortable(fs)) {
       RSSortingTable_Add(&sp->sortables, fs->name, fieldTypeToValueType(fs->types));
     }
-    if (FieldSpec_HasSuffixTrie(fs)) {
+    if (FieldSpec_HasSuffixTrie(fs) && FIELD_IS(fs, INDEXFLD_T_FULLTEXT)) {
```
Comment (raz-mon) [inline src/spec.c:2384]: Actually not covered anyways, but still should happen only for TEXT fields.

File: src/suffix.c [hunk 3]
```diff
@@ -403,13 +396,9 @@ void deleteSuffixTrieMap(TrieMap *trie, const char *str, uint32_t len) {
...
     suffixData *data = TrieMap_Find(trie, str + j, len - j);
-    // suffix trie is shared between all tag fields in index, even if they don't use it.
-    // if the trie is owned by other fields and not any one containing this suffix,
-    // then failure to find the suffix is not an error. just move along.
-    if (data == TRIEMAP_NOTFOUND) continue;
-    // RS_LOG_ASSERT(data != TRIEMAP_NOTFOUND, "all suffixes must exist");
+    RS_LOG_ASSERT(data != TRIEMAP_NOTFOUND, "all suffixes must exist");
     if (j == 0) {
-      // keep pointer to word string to free after it was found in al sub tokens.
+      // keep pointer to word string to free after it was found in all sub tokens.
```
Comment (raz-mon) [inline src/suffix.c:401]: This reverts the changes done in #3289 which are relevant for the TEXT index, but not for the TAG.

File: tests/pytests/test_resp3.py
Comment (nafraf) [inline tests/pytests/test_resp3.py:1]: does worth to add a test creating data, check `total_index_memory_sz_mb`/`text_overhead_sz_mb`/`tag_overhead_sz_mb`, delete data, check them again? WDYT?

**Review by nafraf — COMMENTED** [review]
Nice work. Could we add some test to test `FT.INFO` after creating and deleting data?

File: tests/pytests/test_resp3.py
Comment (raz-mon) [reply tests/pytests/test_resp3.py:1]: Well, they're not `0` after this scenario, so I'm not sure it would help a lot, and also may be flaky.
I'll take another look whether we can add some test for these values, it can be nice.

Comment (raz-mon) [reply tests/pytests/test_resp3.py:1]: @nafraf Added

File: src/suffix.c [hunk 4]
```diff
@@ -51,7 +44,7 @@ void addSuffixTrie(Trie *trie, const char *str, uint32_t len) {
 
   TrieNode *trienode = TrieNode_Get(trie->root, runes, rlen, 1, NULL);
   suffixData *data = NULL;
-  if (trienode && trienode->payload) {
+  if (trienode) {
```
Comment (oshadmi) [inline src/suffix.c:47]: Please explain this change

File: src/suffix.c (hunk 3 above)
Comment (oshadmi) [reply src/suffix.c:401]: I suggest to add a test such as `testSuffixCleanup` from #3289 with a `TAG` field with `WITHSUFFIXTRIE`

Comment (oshadmi) [reply src/suffix.c:401]: Why not have this change and the related test in a separate PR?

File: src/tag_index.c [hunk 5]
```diff
@@ -424,3 +424,44 @@ int TagIndex_RegisterType(RedisModuleCtx *ctx) {
 
   return REDISMODULE_OK;
 }
+
+// static size_t TagIndex_GetSuffixOverhead(TagIndex *idx) {
```
Comment (oshadmi) [inline src/tag_index.c:428]: Can be in separate PR?

File: src/tag_index.c [hunk 6]
```diff
@@ -424,3 +424,44 @@ int TagIndex_RegisterType(RedisModuleCtx *ctx) {
...
+  TagIndex *idx = NULL;
+  RedisSearchCtx sctx = SEARCH_CTX_STATIC(RSDummyContext, sp);
+  RedisModuleString *keyName = TagIndex_FormatName(&sctx, fs->name);
+  idx = TagIndex_Open(&sctx, keyName, 0, NULL);
+  RedisModule_FreeString(RSDummyContext, keyName);
+  if (idx) {
+    overhead = TrieMap_MemUsage(idx->values);     // Values' size are counted in stats.invertedSize
+    if (idx->suffix) {
+      overhead += TrieMap_MemUsage(idx->suffix);
+      // We currently avoid counting the suffix values' overhead for performance.
```
Comment (oshadmi) [inline src/tag_index.c:461]: Can be in separate PR?

**Review by oshadmi — COMMENTED** [review]
Nice! Few comments

File: src/query.c [hunk 7]
```diff
@@ -613,6 +613,10 @@ static IndexIterator *Query_EvalPrefixNode(QueryEvalCtx *q, QueryNode *qn) {
   if (!ctx.its || ctx.nits == 0) {
     rm_free(ctx.its);
     return NULL;
+  // TODO: This should be a single iterator.
```
[the remark this answers is not in the document]
Comment (raz-mon) [reply src/query.c:616]: I think we should keep the TODO, but yes this won't be done in this PR.

File: src/tag_index.c (hunk 5 above)
Comment (raz-mon) [reply src/tag_index.c:428]: What do you mean? I left it as a comment so it won't be counted by coverage, and it can be used for debugging maybe.. I can remove it, but it may be useful for someone someday. I don't understand your meaning with a separate PR.

File: src/tag_index.c (hunk 6 above)
Comment (raz-mon) [reply src/tag_index.c:461]: Why separate PR? This is used.
Do you mean the comment??

File: tests/pytests/test_resp3.py [hunk 8]
```diff
@@ -386,6 +386,9 @@ def test_info():
       'indexing': 0,
       'inverted_sz_mb': ANY,
       'key_table_size_mb': ANY,
+      'tag_overhead_sz_mb': ANY,
+      'text_overhead_sz_mb': ANY,
+      'total_index_memory_sz_mb': ANY,
```
[the remark this answers is not in the document]
Comment (raz-mon) [reply tests/pytests/test_resp3.py:391]: Will add one for it 👍

File: src/tag_index.c (hunk 5 above)
Comment (oshadmi) [reply src/tag_index.c:428]: Let's put it in a PR related to MOD-6865

File: tests/cpptests/json.cpp
[the remark this answers is not in the document]
Comment (raz-mon) [reply tests/cpptests/json.cpp:1]: This file was removed in this PR (that's what the red `minus` sigh says).

**Review by oshadmi — APPROVED** [review]
👏🏼 👍🏼 ❤️
